packages/components/src/input-control/state.js

```diff
@@ -0,0 +1,246 @@
+/**
+ * WordPress dependencies
+ */
+import { useReducer } from '@wordpress/element';
+
+const initialStateReducer = ( state ) => state;
+
+const initialInputControlState = {
+	_event: {},
+	error: null,
+	initialValue: '',
+	isDirty: false,
+	isDragEnabled: false,
+	isDragging: false,
+	isPressEnterToChange: false,
+	value: '',
+};
+
+const actionTypes = {
+	CHANGE: 'CHANGE',
+	COMMIT: 'COMMIT',
+	DRAG_END: 'DRAG_END',
+	DRAG_START: 'DRAG_START',
+	DRAG: 'DRAG',
+	INVALIDATE: 'INVALIDATE',
+	PRESS_DOWN: 'PRESS_DOWN',
+	PRESS_ENTER: 'PRESS_ENTER',
+	PRESS_UP: 'PRESS_UP',
+	RESET: 'RESET',
+	UPDATE: 'UPDATE',
+};
+
+export const inputControlActionTypes = actionTypes;
+
+/**
+ * Prepares initialState for the reducer.
+ *
+ * @param {Object} initialState The initial state.
+ * @return {Object} Prepared initialState for the reducer
+ */
+function mergeInitialState( initialState = initialInputControlState ) {
+	const { value } = initialState;
+
+	return {
+		...initialInputControlState,
+		...initialState,
+		initialValue: value,
+	};
+}
+
+/**
+ * Combines multiple stateReducers into a single stateReducer, building
+ * the pipeline to control the flow for state and actions.
+ *
+ * @param  {...Function} fns State reducers.
+ * @return {Function} The single combined stateReducer.
+ */
+export const composeStateReducers = ( ...fns ) => {
+	return ( ...args ) => {
+		return fns.reduceRight( ( state, fn ) => {
+			return { ...state, ...fn( ...args ) };
```

Comment:
Consider optimising this by making sure that `fn( ...args )` actually returns a non-empty object before concatenating with `state` in a new object. Otherwise, we can avoid returning new references. Something like:

```js
const fnState = fn( ...args );
return isEmpty( fnState ) ? state : { ...state, ...fnState };
```

Whether you'd like to tackle this now or as a separate task, it's your call. :) It may also depend on actual usage of `composeStateReducers`.

Reply:
I think this makes sense! Thank you for your suggestion! I just tested, and it seems to work as expected 👍